tests/test_visualization/test_hierarchical_heatmap.py

```diff
@@ -1,98 +1,126 @@
-        )
-        self.assertIsInstance(mean_intensity, pd.DataFrame)
-        self.assertIsInstance(matrixplot, MatrixPlot)
-
-    def test_invalid_annotation_raises_error(self):
-        """Test if invalid annotation raises error."""
-        err_msg = (r"The annotation 'invalid_annotation' does not exist "
-                   r"in the provided AnnData object. Available annotations "
-                   r"are: \['phenotype'\]")
-        with self.assertRaisesRegex(KeyError, err_msg):
-            hierarchical_heatmap(self.adata, 'invalid_annotation')
-
     def test_nan_annotation_raises_error(self):
         """Test if NaN annotation raises error."""
         self.adata.obs['phenotype'] = [
             None,
             'phenotype1',
             'phenotype2',
-            'phenotype1'
+            'phenotype1',
+            'phenotype3',
+            'phenotype3',
+            'phenotype4',
+            'phenotype4'
         ]
         with self.assertRaises(ValueError):
             hierarchical_heatmap(self.adata, 'phenotype')
 
-    def test_invalid_layer_raises_error(self):
-        """Test if invalid layer raises error."""
-        err_msg = ("The layer 'invalid_layer' does not exist in the "
-                   "provided AnnData object")
-        with self.assertRaisesRegex(KeyError, err_msg):
-            hierarchical_heatmap(
-                self.adata,
-                'phenotype',
-                layer='invalid_layer'
-            )
-
     def test_calculates_correct_mean_intensity(self):
         """Test if correct mean intensities are calculated."""
-        expected_mean_intensity = self.adata.to_df().groupby(
-            self.adata.obs['phenotype']
-        ).mean().reset_index()
-        mean_intensity, _ = hierarchical_heatmap(self.adata, 'phenotype')
-        self.assertEqual(mean_intensity.shape, (2, 3))
+
+        # Ensure annotation column is categorical
+        mean_intensity, _, _ = hierarchical_heatmap(self.adata, 'phenotype')
+        self.assertTrue(
+            pd.api.types.is_categorical_dtype(self.adata.obs['phenotype'])
+        )
+
+        # Calculate the expected mean intensity
+        intensities = self.adata.to_df()
+        labels = self.adata.obs['phenotype']
+        expected_mean_intensity = (
+            pd.concat([intensities, labels], axis=1)
+            .groupby('phenotype')
+            .mean()
+        )
+
+        # Ensure the calculated mean intensity matches the expected
+        # mean intensity
         pd.testing.assert_frame_equal(
-            mean_intensity,
-            expected_mean_intensity,
-            check_exact=True
+            mean_intensity, expected_mean_intensity, check_exact=True
+        )
+
+    def test_zscore_normalization(self):
+        """Test z-score normalization."""
+        _, clustergrid, _ = hierarchical_heatmap(
+            self.adata, 'phenotype', z_score=1
+        )
+        self.assertTrue((clustergrid.data2d.mean().round(2) == 0).all())
```

Comment:
@fangliu117, would you please elaborate why this assertion should be true regarding the returned data? Please add a comment in the code.

Reply:
Thanks, The z-score normalization is to standardize the features such that they have a mean of 
0 and a standard deviation of 1. The formula for z-score normalization is: z= (x−μ)/σ, where x is the original feature value, μ is the mean of the feature, and σ is the standard deviation of the feature.

Given the z-score normalization, the mean of the normalized data for each feature should be approximately 
0. We round to 2 decimal places to account for potential floating-point inaccuracies. This is added to the comments.

Reply:
Thanks @fangliu117